tools/xdp/s2n-quic-xdp/src/syscall.rs

```diff
@@ -0,0 +1,342 @@
+}
+
+/// Opens a mmap region, with an optional associated file descriptor.
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L273).
+#[inline]
+pub fn mmap(len: usize, offset: usize, fd: Option<RawFd>) -> Result<NonNull<c_void>> {
+    unsafe {
+        let flags = if fd.is_some() {
+            libc::MAP_SHARED | libc::MAP_POPULATE
+        } else {
+            libc::MAP_PRIVATE | libc::MAP_ANONYMOUS
+        };
+
+        // See:
+        // * Fill https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L273
+        // * Completion https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L287
+        // * RX https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L1111
+        // * TX https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L1132
+        let addr = libc::mmap(
+            core::ptr::null_mut(),
+            len as _,
+            libc::PROT_READ | libc::PROT_WRITE,
+            flags,
+            fd.unwrap_or(-1),
+            offset as _,
+        );
+
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        let addr = NonNull::new(addr)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "mmap returned null pointer"))?;
+
+        Ok(addr)
+    }
+}
+
+/// Unmaps a mmap region
+///
+/// # Safety
+///
+/// The following must be true:
+///
+/// * The provided `addr` MUST be a mmap'd region
+/// * The original mmap length MUST match the provided `len` parameter.
+pub unsafe fn munmap(addr: NonNull<c_void>, len: usize) -> Result<()> {
+    libc!(munmap(addr.as_ptr(), len as _))?;
+    Ok(())
+}
+
+#[inline]
+fn xdp_option<Fd: AsRawFd, T: Sized>(fd: &Fd, opt: SocketOptions, value: &mut T) -> Result<usize> {
+    let mut optlen = size_of::<T>() as libc::socklen_t;
+
+    libc!(getsockopt(
+        fd.as_raw_fd(),
+        SOL_XDP,
+        opt as _,
+        value as *mut _ as _,
+        &mut optlen,
+    ))?;
+
+    Ok(optlen as usize)
+}
+
+#[inline]
+fn set_xdp_option<Fd: AsRawFd, T: Sized>(fd: &Fd, opt: SocketOptions, value: &T) -> Result<()> {
+    let optlen = size_of::<T>() as libc::socklen_t;
+
+    libc!(setsockopt(
+        fd.as_raw_fd(),
+        SOL_XDP,
+        opt as _,
+        value as *const _ as _,
+        optlen,
+    ))?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::mmap::Mmap;
+    use core::ffi::CStr;
+
+    #[test]
+    fn syscall_test() {
+        // This call requires `CAP_NET_RAW`. If the test doesn't have this set, then log and skip
+        // the next calls
+        let fd = if let Ok(fd) = open() {
+            fd
+        } else {
+            assert!(
+                std::env::var("CAP_NET_RAW_ENABLED").is_err(),
```

Comment:
This is referring to CAP_NET_RAW? Can you prefix CAP_NET_RAW_ENABLED with S2_QUIC

Reply:
It's only used in testing so I'm not sure it makes sense to prefix it.